neqo-transport/src/connection.rs

```diff
@@ -1065,74 +1197,68 @@ impl Connection {
                 State::Closed { .. } => unimplemented!(),
             }
 
+            assert!(encoder.len() <= path.mtu());
             if encoder.len() == 0 {
                 continue;
             }
 
-            qdebug!([self] "Need to send a packet");
+            qdebug!("Need to send a packet");
             match epoch {
                 // Packets containing Initial packets need padding.
                 0 => needs_padding = true,
                 1 => (),
                 // ...unless they include higher epochs.
                 _ => needs_padding = false,
             }
-            let hdr = PacketHdr::new(
-                0,
-                match epoch {
-                    0 => {
-                        let token = match &self.retry_info {
-                            Some(v) => v.token.clone(),
-                            _ => Vec::new(),
-                        };
-                        PacketType::Initial(token)
-                    }
-                    1 => {
-                        assert!(self.zero_rtt_state != ZeroRttState::Rejected);
-                        self.zero_rtt_state = ZeroRttState::Sending;
-                        PacketType::ZeroRTT
-                    }
-                    2 => PacketType::Handshake,
-                    3 => PacketType::Short,
-                    _ => unimplemented!(), // TODO(ekr@rtfm.com): Key Update.
-                },
-                Some(self.version),
-                path.remote_cid.clone(),
-                path.local_cids.first().cloned(),
-                self.loss_recovery.next_pn(space),
-                epoch,
-            );
+
             self.stats.packets_tx += 1;
+            self.loss_recovery.inc_pn(space);
+
+            let mut packet = encode_packet(tx, &hdr, &encoder);
 
-            if ack_eliciting {
-                self.idle_timeout.on_packet_sent(now);
+            if self.output_mode != OutputMode::ProbeTimeout {
+                if ack_eliciting {
+                    self.idle_timeout.on_packet_sent(now);
+                }
+                self.loss_recovery.on_packet_sent(
+                    space,
+                    hdr.pn,
+                    ack_eliciting,
+                    tokens,
+                    packet.len(),
+                    ack_eliciting || has_padding,
```

Comment:
"counts towards cc in-flight" is true if packet is ack_eliciting OR if it contains padding. We already need an ack_eliciting bool for other purposes, so I think it's cleaner to have another bool for contains padding, and then combine them, rather than having one bool for ack_eliciting and another bool for ack_eliciting OR has_padding.